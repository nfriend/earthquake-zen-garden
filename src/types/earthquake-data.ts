// Provides type safety for a subset of the properties in data.json

export interface EarthquakeData {
  profile: {
    firstName: string;
    lastName: string;
    phone: string;
    email: string;
    bio: string;
    avatarImage: string;
  };
  data: {
    features: EarthquakeDataFeature[];
  };
}

export interface EarthquakeDataFeature {
  properties: {
    title: string;
    time: number;
    mag: number;
    status: 'automatic' | 'reviewed';
    tsunami: number;
    type: string;
  };
  id: string;
}
