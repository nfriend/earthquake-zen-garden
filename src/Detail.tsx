import React, { Fragment } from 'react';
import { useParams } from 'react-router';
import { formatDate } from './helpers/format-date';
import PageTitle from './PageTitle';
import { EarthquakeData } from './types/earthquake-data';

const { data }: EarthquakeData = require('./data.json');

function Detail(): JSX.Element {
  const { id } = useParams<{ id: string }>();

  const earthquake = data.features.find((f) => f.id === id);

  const pageTitle = earthquake
    ? earthquake.properties.title
    : 'Earthquake not found!';

  const earthquakeDetails: { label: string; value: string }[] = [
    { label: 'Title', value: earthquake.properties.title },
    { label: 'Magnitude', value: earthquake.properties.mag.toFixed(2) },
    { label: 'Time', value: formatDate(earthquake.properties.time) },
    { label: 'Status', value: earthquake.properties.status },
    { label: 'Tsunami', value: earthquake.properties.tsunami.toString() },
    { label: 'Type', value: earthquake.properties.type },
  ];

  return (
    <div className="flex flex-col items-center">
      <PageTitle title={pageTitle} />
      {earthquake && (
        <div
          className="grid grid-cols-3 max-w-md mx-4"
          data-testid="earthquake-details"
        >
          {earthquakeDetails.map(({ label, value }) => {
            return (
              <Fragment key={label}>
                <h3 className="font-bold mr-2">{label}</h3>
                <span className="col-span-2">{value}</span>
              </Fragment>
            );
          })}
        </div>
      )}
    </div>
  );
}

export default Detail;
