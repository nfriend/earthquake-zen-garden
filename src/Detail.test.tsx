import React from 'react';
import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';
import { Route, MemoryRouter } from 'react-router-dom';
import Detail from './Detail';
import { EarthquakeData } from './types/earthquake-data';

const { data }: EarthquakeData = require('./data.json');

describe('Detail.tsx', () => {
  const renderComponent = (idParam: string) => {
    render(
      <MemoryRouter initialEntries={[`earthquakes/${idParam}`]}>
        <Route path="earthquakes/:id">
          <Detail />
        </Route>
      </MemoryRouter>,
    );
  };

  describe('when the provided earthquake ID is valid', () => {
    beforeEach(() => {
      renderComponent(data.features[0].id);
    });

    test('renders profile details', () => {
      const earthquakeDetails = screen.getByTestId('earthquake-details');
      expect(earthquakeDetails).toMatchSnapshot();
    });
  });

  describe('when the provided earthquake ID is not valid', () => {
    beforeEach(() => {
      renderComponent('not a valid earthquake id');
    });

    test('renders a "not found" message in the page title', () => {
      const titleElement = screen.getByText('Earthquake not found!');
      expect(titleElement).toBeInTheDocument();
    });
  });
});
