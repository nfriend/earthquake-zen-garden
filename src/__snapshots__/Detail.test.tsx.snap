// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Detail.tsx when the provided earthquake ID is valid renders profile details 1`] = `
<div
  class="grid grid-cols-3 max-w-md mx-4"
  data-testid="earthquake-details"
>
  <h3
    class="font-bold mr-2"
  >
    Title
  </h3>
  <span
    class="col-span-2"
  >
    M 1.2 - 3km ENE of The Geysers, CA
  </span>
  <h3
    class="font-bold mr-2"
  >
    Magnitude
  </h3>
  <span
    class="col-span-2"
  >
    1.19
  </span>
  <h3
    class="font-bold mr-2"
  >
    Time
  </h3>
  <span
    class="col-span-2"
  >
    Apr 13, 2018, 07:25 PM
  </span>
  <h3
    class="font-bold mr-2"
  >
    Status
  </h3>
  <span
    class="col-span-2"
  >
    automatic
  </span>
  <h3
    class="font-bold mr-2"
  >
    Tsunami
  </h3>
  <span
    class="col-span-2"
  >
    0
  </span>
  <h3
    class="font-bold mr-2"
  >
    Type
  </h3>
  <span
    class="col-span-2"
  >
    earthquake
  </span>
</div>
`;
